activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb

```diff
@@ -639,6 +639,14 @@ def build_insert_sql(insert) # :nodoc:
         sql
       end
 
+      def assign_next_value_if_serial_column_is_nil!(columns, sequence_name, attributes_array) # :nodoc:
+        serial_column = columns.find(&:serial?)&.name
+        return if serial_column.nil? || sequence_name.nil?
+        attributes_array.each do |attributes|
+          attributes[serial_column] = Arel::Nodes::SqlLiteral.new("nextval('#{sequence_name}')") if attributes[serial_column].nil?
+        end
+      end
```

Comment:
In the case of postgresql only, if the column of serial type (auto increment) is nil, the following error occurs, so a special handling is added.
```
ActiveRecord::NotNullViolation: PG::NotNullViolation: ERROR:  null value in column "id" of relation "books" violates not-null constraint
```